### PR TITLE
[V2] Audio voice fix for channels with user limit

### LINKED ISSUE #1643: [V2] Audio voice
# Other bugs

#### What were you trying to do?

Trying to play music while bot is idle in a voice channel will eror out if that voice channel is full with the bot in the channel. 

#### What were you expecting to happen?

For the bot to keep playing.

#### What actually happened?

Bot errors out with "Your voice channel is full."

#### How can we reproduce this issue?
1. Create a voice channel with 2 slots.
2. Join voice channel and [p]play rickroll
3. After bot has joined and song has finished or you can do [p]stop try to
4. Try to play again while it is you and bot in the VC and channel is 2/2
5. Error

### PR DESCRIPTION
**Pull request by Jackenmen**
# Bugfix request

#### Describe the bug being fixed

When the bot is idle in full channel with user limit, when you were trying to play music with command `[p]play (name of song)`, it was saying "Your voice channel is full."
Fixes #1643

#### Anything we need to know about this fix?
No, all I did was just adding one if (well changing place of it technically). You used such solution in another function already even, it looks like you forgot to do this here as well.

### CODE REVIEW

File: cogs/audio.py [hunk 1]
```diff
@@ -1491,26 +1491,26 @@ def _clean_url(self, url):
...
-            return
-        except UnauthorizedSpeak:
-            await self.bot.say("I don't have permissions to speak in your"
-                               " voice channel.")
-            return
-        except ChannelUserLimit:
-            await self.bot.say("Your voice channel is full.")
-            return
-
         if not self.voice_connected(server):
```
Comment (mikeshardmind) [inline cogs/audio.py:1494]: This is not the correct condition, or at least not without additional conditions, as the else condition prior existing that you left in place would also pass this condition.

File: cogs/audio.py [hunk 2]
```diff
@@ -1491,26 +1491,26 @@ def _clean_url(self, url):
...
-            await self._join_voice_channel(voice_channel)
+            try:
+                self.has_connect_perm(author, server)
```
Comment (mikeshardmind) [inline cogs/audio.py:1496]: This line right here (well, specifically the full channel comparison in the function call (https://github.com/Cog-Creators/Red-DiscordBot/blob/97b70c35ee540710d7172e4fa7f6ac3663a979ea/cogs/audio.py#L2246) is the root cause of the issue, as there is no offset if the bot is in the channel, nor an additional pass condition of the bot is in the channel already with speak perms.

File: cogs/audio.py [hunk 3]
```diff
@@ -1491,26 +1491,26 @@ def _clean_url(self, url):
...
+            except UnauthorizedSpeak:
+                await self.bot.say("I don't have permissions to speak in your"
+                                   " voice channel.")
+                return
+            except ChannelUserLimit:
+                await self.bot.say("Your voice channel is full.")
+                return
+            else:
+                await self._join_voice_channel(voice_channel)
         else:  # We are connected but not to the right channel
```
Comment (mikeshardmind) [inline cogs/audio.py:1514]: Having the else block handled like this means the above permission checking never gets called if you want to bot to switch channels.

**Review by mikeshardmind — CHANGES_REQUESTED** [review]
I think it makes sense to either modify `has_connect_perms` to account for the case of already being in the channel with voice perms (As an admin even join after the bot surpassing the channel limit, making calculating an offest wrong too)

or to modify handling of 

```
except ChannelUserLimit:
```
and handle it there.

I've detailed why the current proposed solution is not quite enough in comments attached.

**Comment by Jackenmen** [comment]
You're right, though I didn't change what was already there. I just changed place of the code - try clause is now in if, like in other places of this code.

Of course this can be fixed in another pull request, I don't think I should do changes not relevant to the issue I mentioned.

I can try to help with that though, because you're right, that this should check for connecting perms in else clause as well.

**Pull request closed without merging**